wurst/event/DamageModify.wurst

```diff
@@ -0,0 +1,327 @@
+	return new Damage(this, target, amount, spell, id)..saveInstance()..modify()..deal()
+
+/** Deals triggered damage from unit to target.
+	You can specify id if there's any unique id interactions. */
+public function unit.damage(unit target, real amount, bool spell, int id)
+	new Damage(this, target, amount, spell, id)..modify()..deal()
+
+/** Deals triggered damage from unit to target. */
+public function unit.damage(unit target, real amount, bool spell)
+	new Damage(this, target, amount, spell, 0)..modify()..deal()
+/** Deals triggered spell damage from unit to target. */
+public function unit.damage(unit target, real amount)
+	new Damage(this, target, amount, true, 0)..modify()..deal()
+
+// ConvertAttackType(7) is a hidden attack type with some unique properties. For
+// more information see goo.gl/9k8tn .
+constant ATTACK_TYPE_UNIVERSAL = ConvertAttackType(7)
+constant DAMAGE_TYPE_CHECK_ID = compiletime(ABIL_ID_GEN.next())
+constant CHEAT_DEATH_ID = compiletime(ABIL_ID_GEN.next())
+
+public class Damage
+	integer id
+	unit source
+	unit target
+	bool spell
+	real initial
+	real amount
+
+	/** check this if you want to save target from damage in all cases */
+	protected bool doSaveTarget = false
+	function saveTarget()
+		doSaveTarget = true
+
+	/** Damage instances are automatically destroyed.
+		Use this to protect them if you need that instances somehow. */
+	function saveInstance()
+		wantSave += 1
+
+	/** Releases Damage instance so it can be destroyed. */
+	function releaseInstance()
+		wantSave -= 1
+		if fired and wantSave <= 0
+			destroy this
+
+	protected int wantSave = 0
+	
+	/** stores how much health were added */
+	protected real afterDamage = 0.
+	protected bool killingBlow = false
+	private bool fired = false
+
+	/** applied all modifiers to the damage instance once */
+	function modify()
+		for i in damageModifiers
+			i.run(this)
+	protected function anounce()
+		for i in damageListeners
+			i.run(this)
+
+	/** Returns true if the damage is going to kill or already have killed target */
+	function isKillingBlow() returns bool
+		return killingBlow or (not fired and target.getHP() - .405 < amount)
+		
+	/** Multiplies initial damage with given magnitude */
+	function addMultiplier(real mult)
+		amount += initial * (mult - 1)
+
+	/** Increases damage amount with given amount */
+	function addAmount(real amount)
+		this.amount += amount
+		
+	construct(unit source, unit target, real amount, bool isSpell, int id)
+		this.source = source
+		this.target = target
+		this.initial = amount
+		this.amount = amount
+		spell = isSpell
+		this.id = id
+
+	protected static function processEvent() returns Damage
+		let damage = GetEventDamage()
+		if damage == 0
+			return null
+		let d = new Damage(GetEventDamageSource(), GetTriggerUnit(), (damage < 0 ? -damage : damage), damage < 0, DAMAGE_ID_NONTRIGGERED)
+		d.fired = true
+		return d
+
+	/** applies user-created Damage instance. */
+	function deal()
+		if not fired
```

Comment:
So `deal` just silently doe nothing if it's already been fired? I don't think that's good enough